File: 2015/11/11.cpp
#include <iostream>


bool check_pass(std::string);
std::string find_pass(std::string);


int main (void) {
    std::string problem_input {"hepxcrrq"};
    std::string new_pass_1 {find_pass(problem_input)};
    std::string new_pass_2 {find_pass(new_pass_1)};
    std::cout << new_pass_1 << std::endl << new_pass_2 << std::endl;
}


bool check_pass(std::string password) {
    bool doublet_0 {}, doublet_1 {}, sequence {};
    for (int i {0}; i < password.length(); i++) {
        if (i >= 2) {
            if (password[i-2] == password[i-1]-1 && password[i-1]-1 == password[i]-2) {
                sequence = true;
            }
            if (password[i-2] == password[i-1] && password[i-1] != password[i]) {
                doublet_0 = true;
                continue;
            }
            if (doublet_0) {
                if (
                    (password[i-2] == password[i-1] && password[i-1] != password[i])
                    ||
                    (password[i-1] == password[i] && i == password.length()-1)
                ) {
                    doublet_1 = true;
                }
            }
        }
    }
    return doublet_0 && doublet_1 && sequence;
}

std::string find_pass(std::string password) {
    int i {1};
    bool j {};
    size_t pass_len {password.length()};
    do {
        if (password[pass_len-i] == 'z') {
            password[pass_len-i] = 'a';
            i++;
            j = true;
            continue;
        }
        password[pass_len-i] = password[pass_len-i]+1;
        if (j) {
            i = 1;
            j = false;
        }
    } while (!check_pass(password));
    return password;
}

// 12.6 ms ± 0.7 ms (1000 runs, 10 warmups) - Clang 14.0.6: clang++ -O3 -std=C++20

File: 2015/11/11.py
def check_pass(string, delay=False):
    if delay:
        return False
    doublet_0, doublet_1, sequence = False, False, False
    for i in range(len(string)):
        if i >= 2:
            if ord(string[i-2]) == ord(string[i-1])-1 == ord(string[i])-2:
                sequence = True
            if string[i-2] == string[i-1] and string[i-1] != string[i]:
                doublet_0 = True
                continue
            if doublet_0:
                if (
                    string[i-2] == string[i-1] and string[i-1] != string[i]
                    ) or (
                    string[i-1] == string[i] and i == len(string)-1
                    ):
                        doublet_1 = True
    return doublet_0 and doublet_1 and sequence


def find_pass(password, delay=False):
    i, j = 1, 0
    while not check_pass(password, delay):
        if delay:
            delay = False
        password = list(password)
        if ord(password[-i]) == ord("z"):
            password[-i] = "a"
            password = "".join(password)
            i += 1
            j = True
            continue
        password[-i] = chr(ord(password[-i])+1)
        if j:
            i = 1
            j = False
        password = "".join(password)
    return password


problem_input = 'hepxcrrq'
new_pass_1 = find_pass(problem_input)
new_pass_2 = find_pass(new_pass_1, True)
print(new_pass_1)
print(new_pass_2)

# 2071 ms ±  34 ms (20 runs, 5 warmups) - Python 3.10.6 (conda-forge)